tests/python/python_tests.bzl

```diff
@@ -0,0 +1,253 @@
+
+def _test_default_non_rules_python(env):
+    py = parse_modules(
+        mctx = _mock_mctx(
+            # NOTE @aignas 2024-09-06: the first item in the module_ctx.modules
+            # could be a non-root module, which is the case if the root module
+            # does not make any calls to the extension.
+            _mod(name = "rules_python", toolchain = [_toolchain("3.11")], is_root = False),
+        ),
+    )
+
+    env.expect.that_str(py.default_python_version).equals("3.11")
+    rules_python_toolchain = struct(
+        name = "python_3_11",
+        python_version = "3.11",
+        register_coverage_tool = False,
+    )
+    env.expect.that_collection(py.toolchains).contains_exactly([rules_python_toolchain])
+
+_tests.append(_test_default_non_rules_python)
+
+def _test_default_non_rules_python_ignore_root_user_error(env):
+    py = parse_modules(
+        mctx = _mock_mctx(
+            _mod(
+                name = "my_module",
+                toolchain = [_toolchain("3.12", ignore_root_user_error = True)],
+            ),
+            _mod(name = "rules_python", toolchain = [_toolchain("3.11")]),
+        ),
+    )
+
+    env.expect.that_bool(py.defaults["ignore_root_user_error"]).equals(True)
+    env.expect.that_str(py.default_python_version).equals("3.12")
+
+    my_module_toolchain = struct(
+        name = "python_3_12",
+        python_version = "3.12",
+        register_coverage_tool = False,
+    )
+    rules_python_toolchain = struct(
+        name = "python_3_11",
+        python_version = "3.11",
+        register_coverage_tool = False,
+    )
+    env.expect.that_collection(py.toolchains).contains_exactly([
+        rules_python_toolchain,
+        my_module_toolchain,
+    ]).in_order()
+
+_tests.append(_test_default_non_rules_python_ignore_root_user_error)
+
+def _test_default_non_rules_python_ignore_root_user_error_non_root_module(env):
+    py = parse_modules(
+        mctx = _mock_mctx(
+            _mod(name = "my_module", toolchain = [_toolchain("3.13")]),
+            _mod(name = "some_module", toolchain = [_toolchain("3.12", ignore_root_user_error = True)]),
+            _mod(name = "rules_python", toolchain = [_toolchain("3.11")]),
+        ),
+    )
+
+    env.expect.that_str(py.default_python_version).equals("3.13")
+    env.expect.that_bool(py.defaults["ignore_root_user_error"]).equals(False)
+
+    my_module_toolchain = struct(
+        name = "python_3_13",
+        python_version = "3.13",
+        register_coverage_tool = False,
+    )
+    some_module_toolchain = struct(
+        name = "python_3_12",
+        python_version = "3.12",
+        register_coverage_tool = False,
+    )
+    rules_python_toolchain = struct(
+        name = "python_3_11",
+        python_version = "3.11",
+        register_coverage_tool = False,
+    )
+    env.expect.that_collection(py.toolchains).contains_exactly([
+        some_module_toolchain,
+        rules_python_toolchain,
+        my_module_toolchain,
```

Comment:
shouldn't the order be `[my, some, rules_python]` ?

Reply:
```suggestion
    env.expect.that_collection(py.toolchains).contains_exactly([
        some_module_toolchain,
        rules_python_toolchain,
        my_module_toolchain, # this was the only toolchain, default to that
```